File: 1049-last-stone-weight-ii/1049-last-stone-weight-ii.ts
function lastStoneWeightII(stones: number[]): number {
    if (stones.length == 1) {
        return stones[0];
    }
    
    let sum: number = 0;
    
    stones.forEach(stone => sum += stone);
    
    let target : number = Math.floor(sum / 2);
    
    let dp : number[] = new Array(target + 1).fill(0);
    
    for (let i = 0; i < stones.length; i++) {
        for (let j = target; j >= stones[i]; j--) {
            dp[j] = Math.max(dp[j], dp[j - stones[i]] + stones[i]);
        }
    }
    
    return sum - 2 * dp[target];
};

File: 1049-last-stone-weight-ii/1049-last-stone-weight-ii.py
class Solution:
    def lastStoneWeightII(self, stones: List[int]) -> int:
        n = len(stones)
        totalSum = sum(stones)
        target = totalSum // 2

        dp = [0] * (target + 1)

        for stone in stones:
            for i in range(target, stone - 1, -1):
                dp[i] = max(dp[i], stone + dp[i - stone])

            if dp[target] == target:
                return totalSum - 2 * target

        return totalSum - 2 * dp[target]
